artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java

```diff
@@ -970,36 +970,37 @@ public long getID() {
 
    @Override
    public int durableUp(Message message) {
-      int count = message.durableUp();
-      if (pagingStore != null) {
-         pagingStore.durableUp(message, count);
-      }
-      return count;
+      return message.durableUp();
    }
 
    @Override
    public int durableDown(Message message) {
-      int count = message.durableDown();
-      if (pagingStore != null) {
-         pagingStore.durableDown(message, count);
-      }
-      return count;
+      return message.durableDown();
    }
 
    @Override
-   public void refUp(Message message) {
-      int count = message.refUp();
+   public void refUp(MessageReference messageReference) {
+      int count = messageReference.getMessage().refUp();
+      if (count == 1) {
+         if (messageReference.getOwner() != null) {
+            messageReference.getOwner().addSize(messageReference.getMessageMemoryEstimate());
+         }
+      }
       if (pagingStore != null) {
-         pagingStore.refUp(message, count);
+         pagingStore.refUp(messageReference.getMessage(), count);
       }
-
    }
 
    @Override
-   public void refDown(Message message) {
-      int count = message.refDown();
+   public void refDown(MessageReference messageReference) {
+      int count = messageReference.getMessage().refDown();
+      if (count == 0) {
+         if (messageReference.getOwner() != null) {
+            messageReference.getOwner().addSize(-messageReference.getMessageMemoryEstimate());
```

Comment:
I would keep an else here, and just use this if null?

Some unit tests may break without this?

you're running the tests.. lets see.

Reply:
it seems only notification/management addresses have a null page store, however the reference count need to be independent of any particular queue having a page store

Reply:
I think for those it should just use the one from the queue where you are then.. I think it would make sense.

Reply:
tests went well. just leave this one as is then...